src/Utility.ts

```diff
@@ -87,63 +54,48 @@ export namespace Utility {
                 return resolve(port);
             });
             server.on('error', (err: Error) => {
-                return reject(new Error(err.toString()));
+                return reject(err);
             });
             server.listen(0, '127.0.0.1');
         });
     }
 
-    export async function openFileIfExists(filepath: string): Promise<boolean> {
-        const exists: boolean = await fse.pathExists(filepath);
-        if (exists) {
-            await vscode.window.showTextDocument(vscode.Uri.file(filepath), { preview: false });
-            return true;
-        } else {
-            return false;
-        }
-    }
-
     export async function getServerPort(serverXml: string): Promise<string> | undefined {
         if (!await fse.pathExists(serverXml)) {
             throw new Error(localize('tomcatExt.noserver', 'No tomcat server.'));
         }
         const xml: string = await fse.readFile(serverXml, 'utf8');
         const jsonObj: {} = await parseXml(xml);
-        return getPortFromJson(jsonObj);
+
+        if (!jsonObj || !jsonObj[Constants.SERVER]) {
+            return undefined;
+        }
+
+        let port: string | undefined;
+        const server: {} = jsonObj[Constants.SERVER];
+        const services: {}[] = server[Constants.SERVICE];
+        if (services) {
+            const service: {} = services.find((item: { $: { name: string } }) => item.$.name === Constants.CATALINA);
+            if (service && service[Constants.CONNECTOR]) {
+                const connectors: { $: {} }[] = service[Constants.CONNECTOR];
+                const connector: { $: {} } = connectors.find((item: { $: { protocol: {} } }) =>
+                    (item.$.protocol === undefined || item.$.protocol.toString().startsWith(Constants.HTTP)));
```

Comment:
item.$.protocol must be a string, rather than to convert it use toString(), another scenario is when item.$.protocol is null, it will break the functions.

Reply:
yes, and I will use try catch to replace those continuous checking.